### PR TITLE
Let flat map writer to add IN_MAP checkpoints to sub nodes

### PR DESCRIPTION
**Pull request by sdruzkin**
TL;DR Allow flat map writer to pass IN_MAP stream checkpoint to the value writer, so that they can be added to the value node row indexes.

In presto-orc each ColumnWriter controls all streams that belong to that node, this also includes
the row index stream containing checkpoints for ALL streams belonging that node. 
However, with addition of the flat map writer, each flat map value node gets an extra IN_MAP 
stream controlled by the flat map writer, we somehow need to add IN_MAP stream checkpoints 
to the flat map value node row index. 

W/o Flat Map / Sequence:
```
Stream 1: Type ROW_INDEX, Node:1 (contains checkpoints from streams 2 and 3)
Stream 2: Type DATA, Node:1
Stream 3: Type LENGTH, Node:1
```

With Flat Maps/ Sequence:
```
Stream 1: Type ROW_INDEX, Node:1, Sequence:1 (needs to contain checkpoints for streams 2-4)
Stream 2: Type IN_MAP, Node:1, Sequence:1  <-- going to be controlled by the flat map writer
Stream 3: Type DATA, Node:1, Sequence:1 
Stream 4: Type LENGTH, Node:1, Sequence:1 
```

There are two ways to solve this problem:
1. add a new InMapStream field to all column writers similar to the present stream, this would also require to add a new method to interact with the IN_MAP stream.
2. update getIndexStream signature to allow passing extra checkpoints.
3. update getIndexStream API to return RowIndex/etc objects instead of StreamDataOutput, and inject extra points there

I decided to go with the option number 2 because it involves less API changes. 

I also removed duplicate code used to build row indexes and unified it under the ColumnWriterUtils. I had to add the `toPositionList` to empty `StreamCheckpoint` interface to make various checkpoints implementations usable 
in the new row index builder method.

Test plan:
- existing tests
- Vader approved with 3.5k samples from all regions

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/ColumnWriterUtils.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+     */
+    @SafeVarargs
+    public static List<RowGroupIndex> buildRowGroupIndexes(
+            boolean compressed,
+            List<ColumnStatistics> rowGroupColumnStatistics,
+            Optional<List<? extends StreamCheckpoint>> prependCheckpoints,
+            PresentOutputStream presentStream,
+            ValueOutputStream<? extends StreamCheckpoint>... dataStreams)
+    {
+        ImmutableList.Builder<RowGroupIndex> rowGroupIndexes = ImmutableList.builder();
```
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/writer/ColumnWriterUtils.java:47]: Should we validate all checkpoints has the same length as rowGroupStatistics ?

Comment (arunthirupathi) [reply presto-orc/src/main/java/com/facebook/presto/orc/writer/ColumnWriterUtils.java:47]: Thoughts on using List<List<Checkpoint>> instead of varargs here.

I thought, it would enforce at least one data stream, but Struct field will not have any checkpoints.

**Review by arunthirupathi — APPROVED** [review]
Vader validates DWRF file format. There are changes in ORC decimal and other writers, are the tests good enough for them ?

**Comment by sdruzkin** [comment]
> Vader validates DWRF file format. There are changes in ORC decimal and other writers, are the tests good enough for them ?

OrcTest is very exhausting. Even a tiny change to checkpoints will break the reader.

**Comment by sdruzkin** [comment]
All tests are green. Good 2 merge.
